370-auth-ui/gateway.md

```diff
@@ -0,0 +1,205 @@
+# Gateway
+
+Skygear Gateway is responsible for:
+
+- Resolve the app and the upstream server from the request host.
+- Delegate request authentication to Skygear Auth.
+- Proxy legacy request convention.
+
+## Resolving the app and the upstream server from the request host
+
+### Default domain
+
+The domain label pattern is `https://[<deployment_version_or_gear_name>.]<app_name>.<skygear_cluster_domain>`.
+
+The defined gear name are `accounts` and `assets`.
+
+Examples:
+
+- `https://698d0e9.myapp.skygearapp.com` resolves to the app `myapp` in deployment version `698d0e9`.
+- `https://accounts.myapp.skygearapp.com` resolves to Skygear Auth of the app `myapp`.
+
+### Custom domain
+
+The resolution of custom domain is purely table-driven. The developer is free to assign domain to their app and gears.
+
+## Delegating request authentication to Skygear Auth
+
+Skygear Gateway initiate subrequest to the resolve endpoint of Skygear Auth after resolving the app. The subrequest does not contain body. The gateway must remove `x-skygear-*` headers from the original request before initiating the subrequest. The gateway must copy `x-skygear-*` headers from the response of the subrequest to the original request.
+
+## Authenticating the request
+
+The resolve endpoint `/resolve` looks at `Authorization:` and `Cookie:` to authenticate the request. `Cookie:` has higher precedence. Regardless of the result, the status code is always 200. If the request is anonymous, the response does not contain any `x-skygear-*` headers. Otherwise the response contains the following headers.
+
+## HTTP Headers
+
+### x-skygear-session-valid
+
+Tell whether the session of the original request is valid.
+
+If this header is absent, it means that the original request does not contain any session.
+
+If the value is `true`, it indicates the original request has valid session. More headers will be included.
+
+If the value is `false`, it indicates that the original request has an invalid session.
+
+### x-skygear-session-transport
+
+Tell the session transport used to resolve the session when `x-skygear-session-valid` is present.
```

Comment:
Is this useful? I don't think the backend service would need to be aware of where the token is in the request.

The only case I can think of is to use the cookie/header in another sub-request in backend. However, the backend should already know where to look for the token, so there is no need to read this header.

Reply:
This header is only useful if we allow the upstream server to easily reuse the cookie or authorization header. Currently we do not ban the upstream server from doing this. The proper way is to support client credentials with admin scope in the future.

Let me remove this header from the spec first.